### DISCUSSION TITLE
Inquire about the implementation details of python/cwe-502 in Codeql

### DISCUSSION THREAD

**Discussion by fe1w0** [github/codeql#16891] [Q&A] [answered]
Hi, CodeQL Team!

I hope this message finds you well.
There are some questions about `codeql-main/python/ql/src/Security/CWE-502/UnsafeDeserialization.ql`, regarding implementation details.

As shown in the following ql file, no other source, sink, or flow to are set in the overall query.
- CWE-502/UnsafeDeserialization.ql
```ql
import python
import semmle.python.security.dataflow.UnsafeDeserializationQuery
import UnsafeDeserializationFlow::PathGraph

from UnsafeDeserializationFlow::PathNode source, UnsafeDeserializationFlow::PathNode sink
where UnsafeDeserializationFlow::flowPath(source, sink)
select sink.getNode(), source, sink, "Unsafe deserialization depends on a $@.", source.getNode(),
  "user-provided value"
```
- UnsafeDeserializationQuery.ql
```ql

private module UnsafeDeserializationConfig implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node source) { source instanceof Source }

  predicate isSink(DataFlow::Node sink) { sink instanceof Sink }

  predicate isBarrier(DataFlow::Node node) { node instanceof Sanitizer }
}

/** Global taint-tracking for detecting "code execution from deserialization" vulnerabilities. */
module UnsafeDeserializationFlow = TaintTracking::Global<UnsafeDeserializationConfig>;
```
Therefore, my confusion is that codeql is the first in the internal process, has the pickle.loads parameter as the sink parameter. If yes, can you tell me about the rule file conveniently ? In addition, what are the default source and sink variables in personal curiosity codeql

**Comment by RasmusWL** [comment, marked as answer]
Hi @fe1w0, if you open the file in VS Code with the CodeQL extension installed, you can simply ctrl+click `Source` to see the definition :blush: 

This takes you to the following piece of code:

https://github.com/github/codeql/blob/4bc4e0845dede4439d7e27db95b09e4316593b09/python/ql/lib/semmle/python/security/dataflow/UnsafeDeserializationCustomizations.qll#L19-L54

What it means is that:

* Any RemoteFlowSouce should be a Source
* input to any decoding that may execute input should be a Sink
* Comparisons with string constants are sanitizers

If you search our code for `pickle.loads` you'll find the concrete modeling, where you can see that it will be recognized as a sink:

https://github.com/github/codeql/blob/f3b27d611a20390b25dec8057d222974d70d184e/python/ql/lib/semmle/python/frameworks/Stdlib.qll#L1343-L1357

**Reply by fe1w0** [reply]
Thank you, RasmusWL, for your patient response. My question has been resolved.🌟😀👍
